<a href="https://colab.research.google.com/github/zmarshall4497/dsba/blob/main/Assignment_2_Adv_Predictive_Modeling%20v2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#import data set
import pandas as pd
from google.colab import files 
import io

upload = files.upload()

dfZM = pd.read_csv(io.BytesIO(upload['BostonHousing.csv']))


Saving BostonHousing.csv to BostonHousing.csv


In [2]:
dfZM.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   CRIM       506 non-null    float64
 1   ZN         506 non-null    float64
 2   INDUS      506 non-null    float64
 3   CHAS       506 non-null    int64  
 4   NOX        506 non-null    float64
 5   RM         506 non-null    float64
 6   AGE        506 non-null    float64
 7   DIS        506 non-null    float64
 8   RAD        506 non-null    int64  
 9   TAX        506 non-null    int64  
 10  PTRATIO    506 non-null    float64
 11  LSTAT      506 non-null    float64
 12  MEDV       506 non-null    float64
 13  CAT. MEDV  506 non-null    int64  
dtypes: float64(10), int64(4)
memory usage: 55.5 KB


In [3]:
dfZM.describe(include='all')

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO       LSTAT  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534   12.653063   
std     28.148861    2.105710    8.707259  168.537116    2.164946    7.141062   
min      2.900000    1.129600    1.000000  187.000000   12.600000    1.730000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000    6.950000   
50%     77.500000    3.207450    5.000000  330.000000   19.050000   11.360000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000   16.955000   
max    100.000000   12.126500   24.000000  711.000000   22.000000   37.970000   

             MEDV   CAT. MEDV  
count  506.000000  506.000000  
mean    22.532806    0.166008  
std      9.197104    0.372456  
min      5.000000    0.000000  
25%     17.025000    0.000000  
50%     21.200000    0.000000  
75%     25.000000    0.000000  
max     50.000000    1.000000

In [4]:
dfZM.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

   LSTAT  MEDV  CAT. MEDV  
0   4.98  24.0          0  
1   9.14  21.6          0  
2   4.03  34.7          1  
3   2.94  33.4          1  
4   5.33  36.2          1

In [5]:
#import necessary libraries 
import numpy as np 
import pandas as pd 
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

In [6]:
#remove continuous DV's
dfZM = dfZM.drop(['MEDV'], axis=1)

In [7]:
dfZM.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

   LSTAT  CAT. MEDV  
0   4.98          0  
1   9.14          0  
2   4.03          1  
3   2.94          1  
4   5.33          1

In [8]:
#Creating Independent Var and Dependent Var
X = dfZM.drop(['CAT. MEDV'], axis=1)
y = dfZM['CAT. MEDV']

In [10]:
print(X)

        CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     PTRATIO  LSTAT  
0       15.3   4.98  
1       17.8   9.14  
2       17.8   4.03  
3       18.7   2.94  
4       18.7 

In [11]:
print(y)

0      0
1      0
2      1
3      1
4      1
      ..
501    0
502    0
503    0
504    0
505    0
Name: CAT. MEDV, Length: 506, dtype: int64


In [12]:
X.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
LSTAT      0
dtype: int64

In [13]:
y.isnull().sum()

0

In [14]:
#Train/Test of Data Frame
X_trainZM, X_testZM, y_trainZM, y_testZM = train_test_split(X, y, test_size=0.2, random_state=42)

In [15]:
#Training the Model
rf_modelZM = RandomForestClassifier(random_state=0)
rf_modelZM.fit(X_trainZM, y_trainZM)

RandomForestClassifier(random_state=0)

In [16]:
rf_modelZM.score(X_testZM, y_testZM)

0.9509803921568627

In [17]:
X.shape

(506, 12)

In [18]:
print(rf_modelZM.n_features_in_)

12


In [19]:
from sklearn.model_selection import GridSearchCV
import numpy as np

param = {'n_estimators': [50,100], 'max_features' : range(1,12)}

In [20]:
rf_modelZM2 = RandomForestClassifier(random_state=0)
grid_search = GridSearchCV(estimator=rf_modelZM2, param_grid=param, n_jobs=-1)
grid_search.fit(X_trainZM,y_trainZM)

GridSearchCV(estimator=RandomForestClassifier(random_state=0), n_jobs=-1,
             param_grid={'max_features': range(1, 12),
                         'n_estimators': [50, 100]})

In [21]:
grid_search.best_params_

{'max_features': 3, 'n_estimators': 100}

In [22]:
grid_search.best_score_

0.9727469135802469

In [23]:
grid_search.score(X_testZM, y_testZM)

0.9509803921568627

In [24]:
#Variable Importance Ranking
final_modelZM = grid_search.best_estimator_
feature_imp = pd.Series(final_modelZM.feature_importances_,index=X.columns).sort_values(ascending=False)

feature_imp

RM         0.382712
LSTAT      0.225458
INDUS      0.069583
PTRATIO    0.068702
TAX        0.052521
DIS        0.042677
CRIM       0.039945
ZN         0.036238
NOX        0.032887
AGE        0.023197
RAD        0.020188
CHAS       0.005892
dtype: float64

In [25]:
#get the confusion matrix for the validation dataset
from sklearn.metrics import confusion_matrix
rf_modelZM.fit(X_trainZM, y_trainZM)
y_predZM = rf_modelZM.predict(X_testZM)
conf_matrix = confusion_matrix(y_testZM, y_predZM)
print(conf_matrix)

[[89  2]
 [ 3  8]]


In [26]:
#Anwsering Question 1

1.	Summarize the random forest model results.
-	What is the optimal number of selected features?
-	Please provide the variable importance ranking. Do they make sense to you?
-	Please provide the confusion matrix for the validation dataset **bold text**

The optimal number of selected features is 3.

The variable importance ranking is as follows:

RM         0.382712
LSTAT      0.225458
INDUS      0.069583
PTRATIO    0.068702
TAX        0.052521
DIS        0.042677
CRIM       0.039945
ZN         0.036238
NOX        0.032887
AGE        0.023197
RAD        0.020188
CHAS       0.005892

This does make sense to me. Based off my confusion matrix, my model is accurately predicting a large number of true negatives and true positives...more than false negatives or false positives. Thus, this tells me that the majority of owner-occupied homes do not exceed the median value.

The reason this importance ranking makes sense to me is because it shows RM (rooms per dwelling) as the first, this would leave me to believe that less rooms = lower house price, thus being below the median. You can also see the second variable is LSTAT which represents the % of lower status of the population, also another indicator of lower home prices based off lower status. INDUS being third tells me there is not alot of non-retail business acres driving the lower than median house prices. These all align with what I am seeing from my confusion matrix.

Confusion Matrix:

 [89  2]

 [3   8]

In [27]:
#implementing SVM Based off these results above...
#import necessary SVC packages
from sklearn.svm import SVC

In [28]:
from sklearn.preprocessing import StandardScaler
X=StandardScaler().fit_transform(X)

In [29]:
print (np.mean(X, axis=0))
print(np.std(X,axis=0))

[-1.12338772e-16  7.89881994e-17  2.10635198e-16 -3.51058664e-17
 -1.96592852e-16 -1.08828186e-16 -1.47444639e-16 -8.42540793e-17
 -1.12338772e-16  0.00000000e+00 -4.21270397e-16 -3.08931624e-16]
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


In [30]:
#train the model with a linear kernel
svc_model_linearZM = SVC(kernel='linear')
svc_model_linearZM.fit(X_trainZM,y_trainZM)


SVC(kernel='linear')

In [31]:
#train the model with a radial kernel
svc_model_radialZM = SVC(kernel='rbf')
svc_model_radialZM.fit(X_trainZM,y_trainZM)

SVC()

In [32]:
#comparing scores of linear vs radial
svc_model_linearZM.score(X_testZM,y_testZM)

0.9509803921568627

In [33]:
svc_model_radialZM.score(X_testZM,y_testZM)

0.8921568627450981

In [34]:
#tuning the model with kernels
param_svc = {'C': [0.01,0.1,1,5,10,100], 'gamma': [0.1,0.25,0.5,0.75,1], 'kernel':['rbf','linear']}
svc_model_ZM_2 = SVC()
grid_search_svc = GridSearchCV(estimator=svc_model_ZM_2, param_grid=param_svc, verbose=3,cv=10)
grid_search_svc.fit(X_trainZM, y_trainZM)

Fitting 10 folds for each of 60 candidates, totalling 600 fits
[CV 1/10] END ....C=0.01, gamma=0.1, kernel=rbf;, score=0.829 total time=   0.0s
[CV 2/10] END ....C=0.01, gamma=0.1, kernel=rbf;, score=0.805 total time=   0.0s
[CV 3/10] END ....C=0.01, gamma=0.1, kernel=rbf;, score=0.805 total time=   0.0s
[CV 4/10] END ....C=0.01, gamma=0.1, kernel=rbf;, score=0.805 total time=   0.0s
[CV 5/10] END ....C=0.01, gamma=0.1, kernel=rbf;, score=0.825 total time=   0.0s
[CV 6/10] END ....C=0.01, gamma=0.1, kernel=rbf;, score=0.825 total time=   0.0s
[CV 7/10] END ....C=0.01, gamma=0.1, kernel=rbf;, score=0.825 total time=   0.0s
[CV 8/10] END ....C=0.01, gamma=0.1, kernel=rbf;, score=0.825 total time=   0.0s
[CV 9/10] END ....C=0.01, gamma=0.1, kernel=rbf;, score=0.825 total time=   0.0s
[CV 10/10] END ...C=0.01, gamma=0.1, kernel=rbf;, score=0.825 total time=   0.0s
[CV 1/10] END .C=0.01, gamma=0.1, kernel=linear;, score=0.902 total time=   0.0s
[CV 2/10] END .C=0.01, gamma=0.1, kernel=linea

GridSearchCV(cv=10, estimator=SVC(),
             param_grid={'C': [0.01, 0.1, 1, 5, 10, 100],
                         'gamma': [0.1, 0.25, 0.5, 0.75, 1],
                         'kernel': ['rbf', 'linear']},
             verbose=3)

In [35]:
#Finding the Optimal Parameters
grid_search_svc.best_params_

{'C': 1, 'gamma': 0.1, 'kernel': 'linear'}

In [36]:
grid_search_svc.score(X_testZM,y_testZM)

0.9509803921568627

In [37]:
#confusion matrix for this

final_modelZM_svc = grid_search_svc.best_estimator_
final_modelZM_svc.fit(X_trainZM, y_trainZM)

SVC(C=1, gamma=0.1, kernel='linear')

In [38]:
y_predZM_svc = final_modelZM_svc.predict(X_testZM)
conf_matrix_svc = confusion_matrix(y_testZM,y_predZM_svc)
print(conf_matrix_svc)

[[88  3]
 [ 2  9]]


2.	For SVM, please compare results for two kernel functions, linear vs. radial. Which kernel function does lead to a better model performance in the default setting?

3.	Please tune the SVM model with selected kernel function, and identify best parameters.
-	Please provide the confusion matrix for the validation dataset
-	What is the optimal parameter(s) for the selected kernel function?

4.	Overall, which method (random forest vs. SVM) is performing better?

Anwsers to the following questions:

2. Based off of the results, a linear kernel function is leading to a better model performance based off of the score.

3. The confusion matrix is 88 TN, 3 FP, 2 FN and 9 TP. The optimal parameters for the linear kernel function is C=1 and Gamma = 0.1.

4. The SVM is performing better than the Random Forest because the SVM has fewer False Negatives and higher False Positives.
